lib/common.py

```diff
@@ -677,6 +697,55 @@ def start_background_process(self, cmd):
     def pid_exists(self, pid):
         return self.ssh_with_result(['test', '-d', '/proc/%s' % pid]).returncode == 0
 
+    def execute_script(self, script_contents, simple_output=True):
+        with tempfile.NamedTemporaryFile('w') as f:
+            f.write(script_contents)
+            f.flush()
+            self.scp(f.name, f.name)
+            try:
+                res = self.ssh(['sh', f.name], simple_output=simple_output)
+                return res
+            finally:
+                self.ssh(['rm', '-f', f.name])
+
+    def distro(self):
+        """ Returns the distro name as detected by the guest tools.
+            If the distro name was not detected, the result will be an empty string
+        """
+        script = "eval $(xe-linux-distribution)\n"
+        script += "echo $os_distro\n"
+        return self.execute_script(script)
+
+    def tools_version_dict(self):
+        """ Returns the guest tools version as detected by the guest tools,
+            as a {major:, minor:, micro:, build:} dict. Values are strings.
+        """
+        return parse_xe_dict(self.param_get('PV-drivers-version'))
+
+    def tools_version(self):
+        """ Returns the tools version in the form major.minor.micro-build """
+        version_dict = self.tools_version_dict()
+        return "{major}.{minor}.{micro}-{build}".format(**version_dict)
+
+    def file_exists(self, filepath):
+        '''Test that the file at filepath exists.'''
+        return self.ssh_with_result(['test', '-f', filepath]).returncode == 0
+
+    def detect_package_manager(self):
+        """ Heuristic to determine the package manager on a unix distro """
+        if self.file_exists('/usr/bin/rpm') or self.file_exists('/bin/rpm'):
+            return 'rpm'
+        elif self.file_exists('/usr/bin/apt-get'):
+            return 'apt-get'
+        else:
+            return 'unknown'
```

Comment:
Why returning `unknown` instead of throwing? Also using an enum would be safer.

Reply:
I don't want the caller to be forced to catch the exception. Agreed about an enum.

Reply:
yeah the exception could be raised by the caller if wanted.